pkg/controllers/disruption/disruptionbudgetmapping.go

```diff
@@ -0,0 +1,169 @@
+}
+
+// HaveBudget checks if there is a remaining budget for a particular nodepool and disruption reason
+func (d DisruptionBudgetMapping) HaveBudget(nodePool, disruptionReason string) bool {
+	if _, ok := d.mapping[nodePool]; !ok {
+		return false
+	}
+	return d.getDefaultOrActualBudget(nodePool, disruptionReason) > 0
+}
+
+// Get returns the remaining budget for a particular nodepool and disruption reason, currently just used for testing
+func (d DisruptionBudgetMapping) Get(nodePool, disruptionReason string) int {
+	return d.mapping[nodePool][disruptionReason]
+}
+
+// Put sets or updates the remaining budget for a particular nodepool and disruption reason
+func (d DisruptionBudgetMapping) Put(nodePool, disruptionReason string, remainingBudget int) {
+	if _, ok := d.mapping[nodePool]; !ok {
+		d.mapping[nodePool] = map[string]int{}
+	}
+	d.mapping[nodePool][disruptionReason] = remainingBudget
+}
+
+// Decrement decrements the remaining budget for a particular nodepool and disruption reason by 1
+func (d DisruptionBudgetMapping) Decrement(nodePool, disruptionReason string) {
+	if _, ok := d.mapping[nodePool]; !ok {
+		return
+	}
+	if _, ok := d.mapping[nodePool][disruptionReason]; !ok {
+		disruptionReason = metrics.DefaultReason
+	}
+	if d.mapping[nodePool][disruptionReason] > 0 {
+		d.mapping[nodePool][disruptionReason]--
+	}
+}
+
+// getDefaultOrActualBudget will return the default budget if the reason provided is not specified in the mapping, otherwise it will return the budget present in the mapping.
+func (d DisruptionBudgetMapping) getDefaultOrActualBudget(nodePool, disruptionReason string) int {
+	if reasonBudget, ok := d.mapping[nodePool][disruptionReason]; ok {
+		return reasonBudget
+	}
+	return d.mapping[nodePool][metrics.DefaultReason]
+}
+
+// BuildDisruptionBudgetMapping will build the disruption budget mapping. The disruption budget maps for each disruption reason the number of allowed disruptions for each node pool.
+// We calculate allowed disruptions by taking the max disruptions allowed by disruption reason and subtracting the number of nodes that are already being deleted by that disruption reason.
+//
+//nolint:gocyclo
+func BuildDisruptionBudgetMapping(ctx context.Context, cluster *state.Cluster, clk clock.Clock, kubeClient client.Client, recorder events.Recorder) (DisruptionBudgetMapping, error) {
+	disruptionBudgetMapping := DisruptionBudgetMapping{mapping: map[string]map[string]int{}}
+	nodePoolList := &v1beta1.NodePoolList{}
+	if err := kubeClient.List(ctx, nodePoolList); err != nil {
+		return DisruptionBudgetMapping{}, fmt.Errorf("listing node pools, %w", err)
+	}
+
+	// We want the nodes in  the cluster, to count the number of nodes in each nodepool, and the number of nodes that are being deleted in each nodepool by disruption reason
+	nodes := cluster.Nodes()
+	numNodes := map[string]int{}   // map[nodepool] -> node count in nodepool
+	disrupting := map[string]int{} // map[nodepool] -> nodes undergoing disruption
```

Comment:
cluster.Disrupting() already exists in cluster state. In this case like you mentioned we need them by nodepool. Looking around I did not see any duplication of this logic

Reply:
Also note we need to count the total number of nodes for each nodepool. 

I am wondering if cluster state could maintain that count as well as we add or remove nodes. We could do same for disrupting so that we don't have to recalculate this every time we invoke the disruption reconcile?